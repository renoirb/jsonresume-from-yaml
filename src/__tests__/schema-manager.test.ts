/* tslint:disable:object-literal-sort-keys */

import { SchemaManager } from '..';

const basics = {
  type: 'object',
  additionalProperties: true,
  properties: {
    name: {
      type: 'string',
    },
    label: {
      type: 'string',
      description: 'e.g. Web Developer',
    },
  },
  required: ['name'],
};

const mockingSchema = {
  id: 'WhateverGoes',
  $schema: 'http://json-schema.org/draft-04/schema#',
  title: 'Resume Schema',
  type: 'object',
  additionalProperties: false,
  properties: { basics, bar: {}, foo: {} /* Notice ambitions is not here */ },
};
const schemaString = JSON.stringify(mockingSchema);

describe('SchemaManager constructor', () => {
  let subject: SchemaManager;
  beforeEach(() => {
    subject = new SchemaManager(schemaString);
  });

  it('We can inject our own initial JSONSchema schema', () => {
    expect(subject).toHaveProperty('schemasString', '[' + schemaString + ']');
    expect(subject).toHaveProperty('sections', {
      WhateverGoes: ['basics', 'bar', 'foo'],
    });
    expect(subject).toHaveProperty('schemaNames', ['WhateverGoes']);
  });
});

describe('SchemaManager.parse', () => {
  let subject: SchemaManager;
  beforeEach(() => {
    subject = new SchemaManager(schemaString);
  });

  it('Knows when a section is not in spec', () => {
    const expectedError = 'Invalid top level section "ambitions" provided,\
      it is not part of the schema; only the following are accepted: \
      basics, bar, foo'.replace(
      /\s+/g,
      ' ',
    );
    expect(() => {
      // @ts-ignore
      subject.parse('WhateverGoes', 'ambitions', {});
    }).toThrowError(expectedError);
  });

  it('Returns a JSON string ONLY when valid', () => {
    const basicsSectionYaml = `
# Comments ignored
name: John Doe
label: Software Developer
drivingLicenseClasses:
- 5
    `;
    const parsed = subject.parse('WhateverGoes', 'basics', basicsSectionYaml);
    expect(parsed).toBe(
      // Notice drivingLicenseClasses does not exist, because JSONSchema
      // has additionalProperties to allow that.
      '{"name":"John Doe","label":"Software Developer","drivingLicenseClasses":[5]}',
    );
    expect(typeof parsed).toBe('string');
    expect(JSON.stringify(parsed)).toMatchSnapshot();
  });

  it('Does not enforce required fields, but this seems like a limitation of ZSchema', () => {
    const basicsSectionYaml = `
# Notice there is no required "name" property
label: Software Developer
    `;
    /**
     * Error will look like this
     * ```
     * Invalid top level section "basics" provided, it is not part of the schema;
     * - code: OBJECT_MISSING_REQUIRED_PROPERTY
     *   params:
     *     - name
     *   message: "Missing required property: name"
     *   path: "#/basics"
     *   schemaId: WhateverGoes
     *
     * the contents has inconsistencies with the schema.
     * ```
     */
    expect(() => {
      subject.parse('WhateverGoes', 'basics', basicsSectionYaml);
    }).toThrowError(/Missing required property: name/);
  });
});

describe('SchemaManager.generateTypingsFileContents', () => {
  let subject: SchemaManager;
  beforeEach(() => {
    subject = new SchemaManager(schemaString);
  });

  it('Generates TypeScript typings', async () => {
    const generated = await subject.generateTypingsFileContents('WhateverGoes');
    expect(generated).toMatchSnapshot();
    const jsonSchemasArray = subject.schemas;
    expect(jsonSchemasArray).toMatchSnapshot();
  });
});
