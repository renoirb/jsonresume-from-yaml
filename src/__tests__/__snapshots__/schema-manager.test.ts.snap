// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaManager.generateTypingsFileContents Generates TypeScript typings 1`] = `
"/**
 * JSONResume schema
 *
 * ==================== DO NOT MODIFY IT BY HAND ====================
 * This file was automatically generated by json-schema-to-typescript
 * for JSONResume schema.
 *
 * If you need to change schema, generate another version using
 * SchemaManager.
 *
 * {@link SchemaManager.generateTypingsFileContents}
 */

export interface ResumeSchema {
  basics?: {
    name: string;
    /**
     * e.g. Web Developer
     */
    label?: string;
    [k: string]: any;
  };
  bar?: any;
  foo?: any;
}
"
`;

exports[`SchemaManager.generateTypingsFileContents Generates TypeScript typings 2`] = `
Array [
  Object {
    "$schema": "http://json-schema.org/draft-04/schema#",
    "additionalProperties": false,
    "id": "WhateverGoes",
    "properties": Object {
      "bar": Object {},
      "basics": Object {
        "additionalProperties": true,
        "properties": Object {
          "label": Object {
            "description": "e.g. Web Developer",
            "type": "string",
          },
          "name": Object {
            "type": "string",
          },
        },
        "required": Array [
          "name",
        ],
        "type": "object",
      },
      "foo": Object {},
    },
    "title": "Resume Schema",
    "type": "object",
  },
]
`;

exports[`SchemaManager.parse Returns a JSON string ONLY when valid 1`] = `"\\"{\\\\\\"name\\\\\\":\\\\\\"John Doe\\\\\\",\\\\\\"label\\\\\\":\\\\\\"Software Developer\\\\\\",\\\\\\"drivingLicenseClasses\\\\\\":[5]}\\""`;
